### PR TITLE
API Version preflight check for sonobuoy run

### LINKED ISSUE #330: Error when running on K8s 1.7.x
Description: When trying to run `sonobuoy run`, it errors out with the following message:

```
[centos@ sonobuoy]$ ./sonobuoy run
INFO[0000] created object                                name=heptio-sonobuoy namespace= resource=namespaces
INFO[0000] created object                                name=sonobuoy-serviceaccount namespace=heptio-sonobuoy resource=serviceaccounts
ERRO[0000] error attempting to run sonobuoy: failed to create object: failed to create API resource sonobuoy-serviceaccount: the server could not find the requested resource
```
After running `sonobuoy gen > manifest.yaml` and `kubectl apply -f manifest.yaml` thereafter produced the following error meesage:

```
[centos@ sonobuoy]$ kubectl apply -f manifest.yaml
namespace "heptio-sonobuoy" created
serviceaccount "sonobuoy-serviceaccount" created
configmap "sonobuoy-config-cm" created
configmap "sonobuoy-plugins-cm" created
pod "sonobuoy" created
service "sonobuoy-master" created
Error from server (BadRequest): error when creating "manifest.yaml": ClusterRoleBinding in version "v1" cannot be handled as a ClusterRoleBinding: no kind "ClusterRoleBinding" is registered for version "rbac.authorization.k8s.io/v1"
Error from server (BadRequest): error when creating "manifest.yaml": ClusterRole in version "v1" cannot be handled as a ClusterRole: no kind "ClusterRole" is registered for version "rbac.authorization.k8s.io/v1"
```

Sonobuoy Version: v0.11.0-alpha (built from source)
Kubectl Version: 1.7.4
Kubernetes Version: 1.7.8
Host OS: CentOS
Host OS Version: 7.4
Bootstrapper: kubeadm

### PR DESCRIPTION
**Pull request by liztio**
Closes #330

### CODE REVIEW

File: pkg/client/preflight.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"fmt"
+
+	version "github.com/hashicorp/go-version"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachineryversion "k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/kubernetes"
+)
+
+var preflightChecks = []func(kubernetes.Interface) error{
```
Comment (chuckha) [inline pkg/client/preflight.go]: whoa did gofmt not put a space here? `error{`

File: pkg/client/preflight.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	preflightDNSCheck,
+	preflightVersionCheck,
+}
+
+// PreflightChecks runs all preflight checks in order, returning the first error encountered.
```
Comment (chuckha) [inline pkg/client/preflight.go]: what was the decision behind bailing on the first one and not collecting a list of errors?

File: pkg/client/preflight.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	if len(obj.Items) == 0 {
+		return errors.New("no kube-dns tests found")
+	}
+
+	return nil
+}
+
+var (
+	minimumKubeVersion = version.Must(version.NewVersion("1.8.0"))
```
Comment (chuckha) [inline pkg/client/preflight.go]: should these be consts up top?

File: pkg/client/preflight.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	maximumKubeVersion = version.Must(version.NewVersion("1.11.0"))
```
Comment (chuckha) [inline pkg/client/preflight.go]: 1.10 is the new one coming out and supporting up to a non-existent version is a little confusing, why are we doing that?

File: pkg/client/preflight.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+func preflightVersionCheck(client kubernetes.Interface) error {
+	res, err := client.
+		Discovery().
+		RESTClient().
+		Get().
+		AbsPath("version").
+		Do().
+		Raw()
+	if err != nil {
+		return errors.Wrap(err, "couldn't retrieve version")
```
Comment (chuckha) [inline pkg/client/preflight.go]: what do you think about saying "can't figure out your version so you may encounter errors. we support versions after <minimum version>" and then passing the check? 

I think I prefer to be more flexible with these preflight checks and only bail if we're sure something isn't going to work. Your call though, I can understand that we'd have more success cases if we can guarantee we support the cluster.

File: pkg/client/preflight.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	serverVersion, err := version.NewVersion(versionInfo.String())
+	if err != nil {
+		return errors.Wrap(err, "couldn't parse version string")
+	}
+
+	if serverVersion.LessThan(minimumKubeVersion) {
+		return fmt.Errorf("Minimum kubernetes version is %s, got %s", minimumKubeVersion.String(), versionInfo.String())
+	}
+
+	if serverVersion.GreaterThan(maximumKubeVersion) {
```
Comment (chuckha) [inline pkg/client/preflight.go]: Could we get away with just the minimum? I'm trying to think of why we'd want a maximum and am not coming up with too many good reasons, maybe if k8s gets to 2.0 we might for a time only support up to 1.y -> 1.x and not 2.0? That reason to me doesn't justify having to figure out a maximum version when we bump these numbers. What am I overlooking?

File: pkg/client/preflight.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+var (
+	minimumKubeVersion = version.Must(version.NewVersion("1.8.0"))
+	maximumKubeVersion = version.Must(version.NewVersion("1.11.0"))
+)
+
+func preflightVersionCheck(client kubernetes.Interface) error {
+	res, err := client.
+		Discovery().
```
Comment (timothysc) [inline pkg/client/preflight.go]: why not Discovery().ServerVersion() ?

File: pkg/client/preflight.go (hunk 3 above)
Comment (timothysc) [inline pkg/client/preflight.go]: I'd almost like to put all the version consts into the versions file, also I don' think we need a maximum.

File: pkg/client/preflight.go (hunk 5 above)
Comment (liztio) [reply pkg/client/preflight.go]: We already have a `--skip-preflight` option. I think that's a good enough escape hatch

File: pkg/client/preflight.go (hunk 7 above)
Comment (liztio) [reply pkg/client/preflight.go]: I... did not know about that!

File: pkg/client/preflight.go (hunk 1 above)
Comment (liztio) [reply pkg/client/preflight.go]: it explicitly removes it. Who knew.

File: pkg/client/preflight.go (hunk 3 above)
Comment (liztio) [reply pkg/client/preflight.go]: I try to keep constants near the function that uses them. Maybe that's not a good pattern?

File: pkg/client/preflight.go (hunk 2 above)
Comment (liztio) [reply pkg/client/preflight.go]: I guess it's not obvious to me how to return a list of errors. There's no ErrorList that implements Error() is there?

File: pkg/client/preflight.go (hunk 1 above)
Comment (chuckha) [reply pkg/client/preflight.go]: 🤯

File: pkg/client/preflight.go (hunk 2 above)
Comment (chuckha) [reply pkg/client/preflight.go]: I was thinking []errors a la kubeadm https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/preflight/checks.go

File: pkg/client/preflight.go (hunk 3 above)
Comment (chuckha) [reply pkg/client/preflight.go]: nah that's fine

File: pkg/client/preflight.go (hunk 2 above)
Comment (liztio) [reply pkg/client/preflight.go]: We'd need to thread that all the way back to client.Interface, is the issue

**Pull request merged**